cpp/src/arrow/flight/types.h

```diff
@@ -750,6 +754,190 @@ struct ARROW_FLIGHT_EXPORT CancelFlightInfoRequest {
   static arrow::Result<CancelFlightInfoRequest> Deserialize(std::string_view serialized);
 };
 
+/// \brief Variant supporting all possible value types for {Set,Get}SessionOptions
+using SessionOptionValue = std::variant<std::string, bool, int32_t, int64_t, float,
+                                        double, std::vector<std::string>>;
+
+/// \brief The result of setting a session option.
+enum class SetSessionOptionStatus : int8_t {
+  /// \brief The status of setting the option is unknown.
+  ///
+  /// Servers should avoid using this value (send a NOT_FOUND error if the requested
+  /// query is not known). Clients can retry the request.
+  kUnspecified,
+  // The session option setting completed successfully.
+  kOk,
+  /// \brief The given session option name was an alias for another option name.
+  kOkMapped,
+  /// \brief The given session option name is invalid.
+  kInvalidKey,
+  /// \brief The session option value is invalid.
+  kInvalidValue,
+  /// \brief The session option cannot be set.
+  kError
+};
+std::ostream& operator<<(std::ostream& os, const SetSessionOptionStatus& r);
+
+/// \brief The result of closing a session.
+enum class CloseSessionStatus : int8_t { kUnspecified, kClosed, kClosing, kNotClosable };
+std::ostream& operator<<(std::ostream& os, const CloseSessionStatus& r);
+
+static const char* const SetSessionOptionStatusNames[] = {
```

Comment:
nit: move this to the .cc file?

Reply:
I'd somewhat wanted this exposed to API users, although unfortunately of course one can't add the rest of the window dressing to the C++ enum classes to go along with it.  Ok with moving it if you think that's cleaner: I don't see that there's a concrete use case it just seemed intuitive to have it visible.  Or are you thinking to define in the .cc and leave it declared here?

Reply:
I think if you want to expose it, it would be best to have an explicit toplevel ToString function to call, like what we do here

https://github.com/apache/arrow/blob/563078fb70f7d23e716a2c1c79e96f7409c02f3f/cpp/src/arrow/filesystem/filesystem.h#L46-L48